### DISCUSSION TITLE
Scripting - JInt - Is it possible to show the line where the error was thrown?

### DISCUSSION THREAD

**Discussion by weirdyang** [OrchardCMS/OrchardCore#16357] [Q&A] [answered]
Using the javascript scripting workflow task, is it possible to show which line or variable that causes an error?

**Comment by hishamco** [comment]
This might be a question for Jint repo

/cc @lahma @hyzx86

**Reply by weirdyang** [reply]
Thanks, so I guess i would have to implement my own script task or script evaluator

**Comment by hyzx86** [comment]
Yes, Jint supports this operation

```c#
          var engine = new Engine(options =>
          {
              options.Interop.ExceptionHandler = ex =>
              {
//Determines the type of exception to be handled by Jint based on the exception information. 
                  Console.WriteLine(ex.Message);
                  return true;
              };
           }
```

**Reply by weirdyang** [reply]
Thanks but I think you misunderstood me, I believe that is already done by default, that is the exception recorded when the workflow script task is faulted. I would like to know the line number of the statement or the variable in the script being evaluated when the exception is thrown.


For example, the fault message:
Cannot read property "length" of null 

I was wondering if the logging can include the statement being evaluated or the variable

**Reply by hyzx86** [reply]
If you want to evaluate the variables at execution time, then you may want to refer to Jint's Debug feature. 
Examples can be found in Jint's unit tests

**Reply by weirdyang** [reply]
cool thanks, will check it out, appreciate the quick response!

**Reply by hyzx86** [reply]
>I believe that is already done by default,

This is not the default, probably because sometimes exceptions may contain confidential information.

**Reply by weirdyang** [reply]
I see, I assumed the fault message was the exception

**Reply by hyzx86** [reply]
> I see, I assumed the fault message was the exception

You don't have to do anything, just try catch. . I think in your scene,

**Reply by weirdyang** [reply]
Hmm looking at the script evaluator used by the workflow script task, I don't see where it hides the exception? https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Workflows/Scripting/JavaScriptWorkflowScriptEvaluator.cs

I don't see it in the engine too

https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore/OrchardCore.Scripting.JavaScript/JavaScriptEngine.cs

**Reply by hyzx86** [reply]
What kind of exception did you encounter? Is it possible that this extension function has been processed internally?

**Comment by weirdyang** [comment, marked as answer]
FOr anyone who is trying to do this, wrap the evaluate method in a try catch and catch `JavaScriptException`, it has a `Location` property

I did not need to modify engine options.

```
            try
            {
                workflowContext.LastResult = await _scriptEvaluator.EvaluateAsync(Script, workflowContext, new OutcomeMethodProvider(outcomes));

            }
            catch (JavaScriptException ex)
            {
                // throw custom exception
                throw new InvalidOperationException(string.Format("Error at Line:{0} Column: {1} | {2}",
                    ex.Location.Start.Line,
                    ex.Location.Start.Column,
                    ex.Message), ex);

            }
```

**Reply by hyzx86** [reply]
Yes, but it should need to work with the options above, otherwise it won't throw the exception position in the custom function

**Reply by weirdyang** [reply]
Hmm it seems to be working correctly out of the box though didn't need to use any options. I just modified the default script task without changing any options, happy to be proven otherwise though, let me know if you have a sample to show what you mean
![image](https://github.com/user-attachments/assets/00b067b4-c833-4722-95a0-7a161bf93618)

**Reply by hyzx86** [reply]
Really? As I recall, when an extension reports an error internally, it doesn't indicate the location of the script where the error occurred, e.g. if you make an extension that handles excel, if excel doesn't exist, it will throw an IO exception without including the exact location of the script.

**Comment by weirdyang** [comment]
Yeap I think it’s cos it’s using Jint, and the engine itself is throwing
the exception. So you have to catch the specific exception to case it to
JavascrptException.
